sound/soc/sof/ipc4-debug-stream.c

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2024 Intel Corporation. All rights reserved.
+//
+
+#include <linux/debugfs.h>
+#include <linux/io.h>
+#include <linux/pm_runtime.h>
+#include <sound/sof/debug.h>
+#include <sound/sof/ipc4/header.h>
+#include "sof-priv.h"
+#include "ops.h"
+#include "ipc4-debug-stream.h"
+#include "ipc4-priv.h"
+
+static ssize_t sof_debug_stream_entry_read(struct file *file, char __user *buffer,
+					size_t count, loff_t *ppos)
+{
+	struct snd_sof_dfsentry *dfse = file->private_data;
+	struct snd_sof_dev *sdev = dfse->sdev;
+	u32 type = SOF_IPC4_DEBUG_SLOT_DEBUG_STREAM;
+	loff_t pos = *ppos;
+	size_t size_ret;
+	u32 offset;
+	u8 *buf;
+
+	if (pos < 0)
+		return -EINVAL;
+	if (pos >= SOF_IPC4_DEBUG_SLOT_SIZE || !count)
+		return 0;
+	if (count > SOF_IPC4_DEBUG_SLOT_SIZE - pos)
+		count = SOF_IPC4_DEBUG_SLOT_SIZE - pos;
+
+	offset = sof_ipc4_find_debug_slot_offset_by_type(sdev, type);
+	if (!offset)
+		return -EFAULT;
+
+	buf = kzalloc(SOF_IPC4_DEBUG_SLOT_SIZE, GFP_KERNEL);
+	if (!buf)
+		return -ENOMEM;
+
+	sof_mailbox_read(sdev, offset, buf, SOF_IPC4_DEBUG_SLOT_SIZE);
```

Comment:
Would not make sense to skip the magic number as the telemetry does?
If so, then probably creating a single function to get the content of a debug slot by type and use that in both?

Reply:
Sorry @ujfalusi , I do not follow. What magic number you are referring to?

About unifying the telemetry and debug-stream code I agree on general level, but there is the exception part in the current telemetry code, that I do not fully comprehend. And then telemetry debugfs does not support seek and there are some other subtle differences. To unify the code I should know how telemetry debugfs file is used, so that I can verify it works with code borrowed from debug-stream -side. The other way around wont work, as my python debug-stream client disagrees with the telemetry debugfs implementation.

Reply:
@jsarha, the first u32 in a slot is the magic number or tag or slot type, it is SOF_IPC4_DEBUG_SLOT_DEBUG_STREAM in this case. For telemetry it is SOF_IPC4_DEBUG_SLOT_TELEMETRY.
The real data is after this magic number.

The `sof_ipc4_query_exception_address()` is just wrapping the `sof_ipc4_find_debug_slot_offset_by_type()` call and skips the magic number to get a pointer for the slot data.
The telemetry.c uses memcpy_fromio() instead sof_mailbox_read(), they do the same at the end.

If I look at the two files from a distance, they are 99% identical with the only difference is:
SOF_IPC4_DEBUG_SLOT_DEBUG_STREAM vs SOF_IPC4_DEBUG_SLOT_TELEMETRY
"debug_stream" vs "exception"

The only tricky part is how to know which slot you need to return, for which the `debug-dsp-ops.c` file gives an answer.

Reply:
Yes, I know how the debug slots work, and I understand the basic structure, but while I was implementing debug-stream, I started off with how telemetry debugfs was implemented, but needed to change couple of things  to get it to work.

Only thing is, that I would like to test the telemetry still works, if I unify the debugfs behind a single function that follows the debug-stream implementation... but I have no idea how telemetry should be used and I can not find any client etc. But yes, maybe I just do it and hope the possible issues will come up in review.

Reply:
Right, yes, you are right. The telemetry slot itself can be split up to different sections and internally it is using 'separator magic' numbers for identification (in theory, but that's another story).
So, here, the ipc4-telemetry.c is exposing the 'exception' section of the telemetry slot, which I guess is the only one present when an exception happens and it will overtake the whole telemetry slot.
This is the reason why the first u32 is skipped I guess, it is a marker in the scope of the telemetry slot.